### PR TITLE
Wandering Traders are now serialized properly

### PR DESCRIPTION
**Pull request by Narimm**
Villagers are improved the following is serialized
  - Type
  - Recipes
Entities that can hold equipment = equipment is serialized.
Endermen are serialized and the held block is tracked.

Signed-off-by: Narimm <benjicharlton@gmail.com>

### CODE REVIEW

File: src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java [hunk 1]
```diff
@@ -12,20 +13,31 @@
 import org.bukkit.entity.Tameable;
 import org.bukkit.event.entity.EntityDamageByEntityEvent;
 import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.inventory.EntityEquipment;
+import org.bukkit.inventory.EquipmentSlot;
+import org.bukkit.inventory.Inventory;
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java:18]: Codacy found an issue: [Avoid unused imports such as 'org.bukkit.inventory.Inventory'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Complexity increasing per file
==============================
- src/main/java/me/botsko/prism/actions/entity/WolfSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/ZombieVillageSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/AbstractVillagerSerializer.java  1
- src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java  1
- src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java  4
- src/main/java/me/botsko/prism/actions/entity/EndermanSerializer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169&bid=20664411)

File: src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package me.botsko.prism.actions.entity;
+
+import com.google.common.base.Preconditions;
+import org.bukkit.Material;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.Serializable;
+import javax.annotation.Nonnull;
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java:10]: Codacy found an issue: [Avoid unused imports such as 'javax.annotation.Nonnull'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java [hunk 3]
```diff
@@ -12,20 +13,31 @@
...
+import org.bukkit.inventory.ItemStack;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java:23]: Codacy found an issue: [Avoid unused imports such as 'java.util.function.BiConsumer'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
+package me.botsko.prism.actions.entity;
+
+import com.google.common.base.Preconditions;
+import org.bukkit.Material;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.Serializable;
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java:9]: Codacy found an issue: [Import statement for 'java.io.Serializable' is in the wrong order. Should be in the 'STANDARD_JAVA_PACKAGE' group, expecting group 'SPECIAL_IMPORTS' on this line.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/me/botsko/prism/actions/entity/EndermanSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java  4
- src/main/java/me/botsko/prism/actions/entity/AbstractVillagerSerializer.java  1
- src/main/java/me/botsko/prism/actions/entity/WolfSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/ZombieVillageSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169&bid=20664411)

File: src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java (hunk 1 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java:18]: Codacy found an issue: [Avoid unused imports such as 'org.bukkit.inventory.Inventory'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java (hunk 3 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java:23]: Codacy found an issue: [Avoid unused imports such as 'java.util.function.BiConsumer'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/main/java/me/botsko/prism/actions/entity/WolfSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java  4
- src/main/java/me/botsko/prism/actions/entity/ZombieVillageSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/EndermanSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java  1
- src/main/java/me/botsko/prism/actions/entity/AbstractVillagerSerializer.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169&bid=20664411)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/me/botsko/prism/actions/entity/AbstractVillagerSerializer.java  1
- src/main/java/me/botsko/prism/actions/entity/PrismItemStack.java  1
- src/main/java/me/botsko/prism/actions/entity/ZombieVillageSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/EndermanSerializer.java  2
- src/main/java/me/botsko/prism/actions/entity/EntitySerializer.java  4
- src/main/java/me/botsko/prism/actions/entity/WolfSerializer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169&bid=20664411)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+                if (effect.hasTrail()) {
+                    data.hasTrail = true;
+                }
+            }
+        }
+    }
+
+    public ItemStack toBukkit() {
+        Material m = Material.matchMaterial(material);
+        if (m == null)
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java]: Codacy found an issue: ['if' construct must use '{}'s.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

**Comment by Narimm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 17
           

Complexity increasing per file
==============================
- src/main/java/me/botsko/prism/serializers/entity/PandaSerializer.java  7
- src/main/java/me/botsko/prism/serializers/entity/VillagerSerializer.java  2
- src/main/java/me/botsko/prism/serializers/entity/CatSerializer.java  2
- src/main/java/me/botsko/prism/serializers/entity/SheepSerializer.java  2
- src/main/java/me/botsko/prism/serializers/entity/WolfSerializer.java  2
- src/main/java/me/botsko/prism/serializers/entity/EntitySerializerFactory.java  3
- src/main/java/me/botsko/prism/serializers/entity/AbstractHorseSerializer.java  9
- src/main/java/me/botsko/prism/serializers/entity/ZombieVillageSerializer.java  2
- src/main/java/me/botsko/prism/serializers/entity/EntitySerializer.java  14
- src/main/java/me/botsko/prism/serializers/items/StorageItemStackSerializer.java  3
- src/main/java/me/botsko/prism/serializers/entity/EndermanSerializer.java  2
- src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java  23
- src/main/java/me/botsko/prism/serializers/entity/AbstractVillagerSerializer.java  1
- src/main/java/me/botsko/prism/serializers/entity/ParrotSerializer.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169&bid=20664411)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                    } else {
+                        item.addUnsafeEnchantment(enchantment, Integer.parseInt(enchArgs[1]));
+                    }
+                }
+            }
+        }
+
+        ItemMeta meta = item.getItemMeta();
+        if (name != null) {
+            if (meta != null) {
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java:232]: Codacy found an issue: [These nested if statements could be combined](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    public String slot = "-1";
+    public int[] effectColors;
+    public int[] fadeColors;
+    public boolean hasFlicker;
+    public boolean hasTrail;
+    public short durability = 0;
+    public Map<String, String> bannerMeta;
+    public String material;
+
+    public static ItemStackSerializer createItemStackSerialized(ItemStack item) {
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java:70]: Codacy found an issue: [The method 'createItemStackSerialized(ItemStack)' has an NPath complexity of 32256, current threshold is 200](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+            stringStringMap.forEach((patternIdentifier, dyeName) -> {
+                PatternType type = PatternType.getByIdentifier(patternIdentifier);
+                try {
+                    DyeColor color = DyeColor.valueOf(dyeName);
+                    if (type != null) {
+                        Pattern p = new Pattern(color, type);
+                        patterns.add(p);
+                    }
+                } catch (IllegalArgumentException e){
+                    Prism.debug("Found illegal DyeColor stored in serialized banner: "+e.getMessage());
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java]: Codacy found an issue: [WhitespaceAround: '+' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/entity/VillagerSerializer.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
+package me.botsko.prism.serializers.entity;
+
+import me.botsko.prism.serializers.items.ItemStackSerializer;
+import me.botsko.prism.utils.MiscUtils;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Villager;
+import org.bukkit.entity.Villager.Profession;
+import org.bukkit.inventory.MerchantRecipe;
+
+import java.io.Serializable;
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/entity/VillagerSerializer.java]: Codacy found an issue: [Avoid unused imports such as 'java.io.Serializable'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/entity/AbstractHorseSerializer.java [hunk 10]
```diff
@@ -1,114 +1,114 @@
...
+        h.getAttribute(Attribute.GENERIC_MAX_HEALTH).setBaseValue(maxHealth);
+
+        AttributeInstance attributeInstance = h.getAttribute(Attribute.GENERIC_MOVEMENT_SPEED);
+        if (attributeInstance != null) {
+            attributeInstance.setBaseValue(movementSpeed);
+        }
+    }
+
+    @Override
+    protected void niceName(StringBuilder sb, int start) {
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/entity/AbstractHorseSerializer.java]: Codacy found an issue: [Avoid reassigning parameters such as 'start'](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java (hunk 7 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java:70]: Codacy found an issue: [Missing a Javadoc comment.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/actions/ItemStackAction.java [hunk 11]
```diff
@@ -39,26 +40,36 @@
...
-    private ItemStackActionData actionData;
-    private Map<Enchantment, Integer> enchantments;
+    private ItemStackSerializer actionData;
+
     /**
      * Holds durability if no actionData yet exists.
      */
     private short tempDurability = -1;
 
+    private static ItemStackSerializer createItemStackActionData(ItemStack item, @Nullable  Map<Enchantment, Integer> enchantments) {
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/actions/ItemStackAction.java]: Codacy found an issue: [Line is longer than 120 characters (found 133).](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+            List<Pattern> patterns = new ArrayList<>();
+            stringStringMap.forEach((patternIdentifier, dyeName) -> {
+                PatternType type = PatternType.getByIdentifier(patternIdentifier);
+                try {
+                    DyeColor color = DyeColor.valueOf(dyeName);
+                    if (type != null) {
+                        Pattern p = new Pattern(color, type);
+                        patterns.add(p);
+                    }
+                } catch (IllegalArgumentException e){
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java]: Codacy found an issue: [WhitespaceAround: '{' is not preceded with whitespace.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+            }
+        }
+        if (meta instanceof FireworkEffectMeta) {
+            applyFireWorkMetaToActionData(meta, data);
+        }
+        if (meta instanceof BannerMeta) {
+            List<Pattern> patterns = ((BannerMeta) meta).getPatterns();
+            Map<String, String> stringyPatterns = new HashMap<>();
+            patterns.forEach(
+                    pattern -> stringyPatterns.put(pattern.getPattern().getIdentifier(), pattern.getColor().name()));
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java]: Codacy found an issue: ['lambda arguments' has incorrect indentation level 20, expected level should be one of the following: 16, 18.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+                PatternType type = PatternType.getByIdentifier(patternIdentifier);
+                try {
+                    DyeColor color = DyeColor.valueOf(dyeName);
+                    if (type != null) {
+                        Pattern p = new Pattern(color, type);
+                        patterns.add(p);
+                    }
+                } catch (IllegalArgumentException e){
+                    Prism.debug("Found illegal DyeColor stored in serialized banner: "+e.getMessage());
+                    Prism.debug("Data:  " +bannerMeta.toString());
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java]: Codacy found an issue: [WhitespaceAround: '+' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+                    data.hasFlicker = true;
+                }
+                if (effect.hasTrail()) {
+                    data.hasTrail = true;
+                }
+            }
+        }
+    }
+
+    public ItemStack toBukkit() {
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java:200]: Codacy found an issue: [Missing a Javadoc comment.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/entity/VillagerSerializer.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    protected void serializer(Entity entity) {
+        super.serializer(entity);
+        type = ((Villager) entity).getVillagerType().name().toLowerCase();
+        profession = ((Villager) entity).getProfession().name().toLowerCase();
+        villagerXp = ((Villager) entity).getVillagerExperience();
+        for (MerchantRecipe recipe : ((Villager) entity).getRecipes()) {
+            final VillagerRecipe r = new VillagerRecipe();
+            r.result = ItemStackSerializer.createItemStackSerialized(recipe.getResult());
+            r.ingredient = new ArrayList<>();
+            recipe.getIngredients().forEach(rec -> r.ingredient.add(ItemStackSerializer.createItemStackSerialized(rec)));
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/entity/VillagerSerializer.java]: Codacy found an issue: [Line is longer than 120 characters (found 121).](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+            for (final Map.Entry<Enchantment, Integer> ench : item.getEnchantments().entrySet()) {
+                // This is silly
+                enchs[i] = ench.getKey().getKey().getKey() + ":" + ench.getValue();
+                i++;
+            }
+            data.enchs = enchs;
+        } else if (meta instanceof EnchantmentStorageMeta) {
+            final EnchantmentStorageMeta bookEnchantments = (EnchantmentStorageMeta) meta;
+            if (bookEnchantments.hasStoredEnchants()) {
+                if (bookEnchantments.getStoredEnchants().size() > 0) {
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java:138]: Codacy found an issue: [These nested if statements could be combined](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java (hunk 15 above)
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/ItemStackSerializer.java:200]: Codacy found an issue: [The method 'toBukkit()' has an NPath complexity of 3276, current threshold is 200](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

File: src/main/java/me/botsko/prism/serializers/items/StorageItemStackSerializer.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+public class StorageItemStackSerializer extends ItemStackSerializer {
+    public List<ItemStackSerializer> inventoryContents = new ArrayList<>();
+
+    @Override
+    public ItemStack toBukkit() {
+        //todo
+        return super.toBukkit();
+    }
+
+    public static StorageItemStackSerializer deserialize(String data) {
```
Comment (Narimm) [inline src/main/java/me/botsko/prism/serializers/items/StorageItemStackSerializer.java]: Codacy found an issue: [Missing a Javadoc comment.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=6368169)

**Comment by viveleroi** [comment]
This repository is no longer active. You can try to resubmit this at the current Prism repo here https://github.com/prism/PrismRefracted (athough, some of these older fix PRs were already adapted, please verify first)

**Pull request closed without merging**